samples/jupyter-notebooks/ALS_python_demo.ipynb

```diff
@@ -0,0 +1,639 @@
+    "\n",
+    "m = nrow (R);\n",
+    "n = ncol (R);\n",
+    "\n",
+    "# initializing factor matrices\n",
+    "U = rand (rows = m, cols = r, min = -0.5, max = 0.5);\n",
+    "M = rand (rows = n, cols = r, min = -0.5, max = 0.5);\n",
+    "\n",
+    "# initializing transformed matrices\n",
+    "Rt = t(R);\n",
+    "\n",
+    "\n",
+    "loss_init = 0.0; # only used if check is TRUE\n",
+    "if (check) {\n",
+    "  loss_init = sum (R_nonzero_ind * (R - (U %*% t(M)))^2) + lambda * (sum ((U^2) * row_nonzeros) + sum ((M^2) * col_nonzeros));\n",
+    "  print (\"----- Initial train loss: \" + loss_init + \" -----\");\n",
+    "}\n",
+    "\n",
+    "lambda_I = diag (matrix (lambda, rows = r, cols = 1));\n",
+    "it = 0;\n",
+    "converged = FALSE;\n",
+    "while ((it < max_iter) & (!converged)) {\n",
+    "  it = it + 1;\n",
+    "  # keep M fixed and update U\n",
+    "  parfor (i in 1:m) {\n",
+    "    M_nonzero_ind = t(R[i,] != 0);\n",
+    "    M_nonzero = removeEmpty (target=M * M_nonzero_ind, margin=\"rows\");\n",
+    "    A1 = (t(M_nonzero) %*% M_nonzero) + (as.scalar(row_nonzeros[i,1]) * lambda_I); # coefficient matrix\n",
+    "    U[i,] = t(solve (A1, t(R[i,] %*% M)));\n",
+    "  }\n",
+    "\n",
+    "  # keep U fixed and update M\n",
+    "  parfor (j in 1:n) {\n",
+    "    U_nonzero_ind = t(Rt[j,] != 0)\n",
+    "    U_nonzero = removeEmpty (target=U * U_nonzero_ind, margin=\"rows\");\n",
+    "    A2 = (t(U_nonzero) %*% U_nonzero) + (as.scalar(col_nonzeros[j,1]) * lambda_I); # coefficient matrix\n",
+    "    M[j,] = t(solve (A2, t(Rt[j,] %*% U)));\n",
+    "  }\n",
+    "\n",
+    "  # check for convergence\n",
+    "  if (check) {\n",
+    "    loss_cur = sum (R_nonzero_ind * (R - (U %*% t(M)))^2) + lambda * (sum ((U^2) * row_nonzeros) + sum ((M^2) * col_nonzeros));\n",
+    "    loss_dec = (loss_init - loss_cur) / loss_init;\n",
+    "    print (\"Train loss at iteration (M) \" + it + \": \" + loss_cur + \" loss-dec \" + loss_dec);\n",
+    "    if (loss_dec >= 0 & loss_dec < thr | loss_init == 0) {\n",
+    "      print (\"----- ALS converged after \" + it + \" iterations!\");\n",
+    "      converged = TRUE;\n",
+    "    }\n",
+    "    loss_init = loss_cur;\n",
+    "    print(loss_init)\n",
+    "  }\n",
+    "} # end of while loop\n",
+    "\n",
+    "if (check) {\n",
+    "  print (\"----- Final train loss: \" + loss_init + \" -----\");\n",
+    "}\n",
+    "\n",
+    "if (!converged) {\n",
+    "  print (\"Max iteration achieved but not converged!\");\n",
+    "}\n",
+    "\n",
+    "# inject 0s in U if original R had empty rows\n",
+    "if (num_zero_rows > 0) {\n",
+    "  U = removeEmpty (target = diag (orig_nonzero_rows_ind), margin = \"cols\") %*% U;\n",
+    "}\n",
+    "# inject 0s in R if original V had empty rows\n",
+    "if (num_zero_cols > 0) {\n",
+    "  M = removeEmpty (target = diag (orig_nonzero_cols_ind), margin = \"cols\") %*% M;\n",
+    "}\n",
+    "M = t(M);\n",
+    "\"\"\""
```

Comment:
Can you remove the extra spaces between the function names and the opening parentheses throughout this script?  That issue is fairly prevalent in the older scripts, and it's caused confusion due to being hard to read.

I.e. `rand (rows = m, cols = r, min = -0.5, max = 0.5)` --> `rand(rows = m, cols = r, min = -0.5, max = 0.5)`